editoast: allow users to have multiple identities

Sh099078 opened the pull request:
Fixes https://github.com/OpenRailAssociation/osrd/issues/13822.

Editoast: 
- Update the database model in order to be able to associate several identities with a user. Create a new table `authn_user_identity` which associates users to their identities.
- Handle multiple identities per user in the Editoast CLI user commands when creating or listing users. Add a new user subcommand `add-identities` to add new identities to an existing user.
- Remove user identities from the search criteria and results.

Front: update impersonification to retrieve a user indentity through the `authz/user/info` endpoint instead of getting if from a user search.

CI: docker-compose: update the `editoast add user` in osrd compose to take into account the fact that the order of the identity and name arguments of that command got reversed.

Note: maybe we should also add a `remove-identity` command ? The only way to remove an identity from a user at the moment is to delete and re-create the user.

Khoyo commented on editoast/migrations/2025-11-19-092852-0000_modify_authn_user_to_handle_multiple_identities_for_same_user/up.sql:
```diff
@@ -0,0 +1,11 @@
+create table authn_user_identity (
+    id bigint primary key generated by default as identity,
+    user_id bigint references authn_user on delete cascade not null,
+    identity varchar(255) not null,
```
Why not just a TEXT column? The length limitation seems like a randomly picked magic number, or a misguided belief that it's more performant than TEXT (it's not, it's actually the other way around)

Sh099078 replied:
You're right, that size limit seems arbitrary. It was already a VARCHAR(255) in the prior model and I didn't challenge it but we should indeed change it to TEXT while we are at it. Maybe @leovalais has some insight on it since the original `authn_user` table was created in a commit co-authored by you two. Otherwise I am also in favour of changing to TEXT there.

leovalais replied:
I don't remember why and I can't see a good reason to not have TEXT either.

Sh099078 replied:
Column updated to TEXT, resolving that thread

flomonster commented:
> Note: maybe we should also add a `remove-identity` command ? The only way to remove an identity from a user at the moment is to delete and re-create the user.

We can do this in a follow up PR.

flomonster commented on editoast/migrations/2025-11-19-092852-0000_modify_authn_user_to_handle_multiple_identities_for_same_user/down.sql:
```diff
@@ -0,0 +1,18 @@
+alter table authn_user add column identity_id varchar(255);
+
+-- TODO PR: do we want to save one of the user identities on rollback ?
+--          should we pick it arbitrarily ?
```
Picking arbitrarily on of it's identity seems fine. If it does not have an identity then we can skip the user.

flomonster commented on editoast/src/models/auth_driver.rs:
```diff
@@ -30,6 +32,10 @@ pub enum AuthDriverError {
     #[error(transparent)]
     OpenFgaRequestFailure(#[from] fga::client::RequestFailure),
+    // #[error("At least one of the provided identities is already associated to another user")]
+    // IdentityAlreadyUsed(Vec<String>),
```
We can delete this, not having the list of already associated identities seems fine.

flomonster commented on editoast/editoast_models/src/auth_driver.rs:
```diff
@@ -100,14 +114,31 @@ impl StorageDriver for PgAuthDriver {
+                ON u.id = i.user_id
+                WHERE u.id = $1
+                GROUP BY u.id, u.name;
+            ";
```
Did you try to use diesel instead of a raw query. It's possible to run `group by` queries. Check doc [here](https://docs.diesel.rs/2.0.x/diesel/associations/trait.GroupedBy.html).

Khoyo replied:
I believe diesel chocked on `ARRAY_AGG`.

flomonster replied:
I see, I let you close this conversion in this case.

Sh099078 replied:
> I believe diesel chocked on `ARRAY_AGG`.

Yep, if there is a cleaner way to do it with the dsl I didn't find it

flomonster reviewed (commented):
Thanks for this PR. I still have to test the PR to approve. I left some coments. I also let you fix the CI.

woshilapin commented on editoast/database/src/lib.rs:
```diff
@@ -11,4 +11,4 @@ pub use db_connection_pool::DbConnectionPoolV2;
 #[derive(Debug, thiserror::Error, PartialEq)]
 #[error("an error occurred while querying the database: {0}")]
-pub struct DatabaseError(#[from] diesel::result::Error);
+pub struct DatabaseError(#[from] pub diesel::result::Error);
```
Why do we need the `pub` since we have the `#[from]` to build it? If I understand, at some point, you need to know if the underlying error is a unique violation. If this is the case, we can expose a function `DatabaseError::is_unique_violation`. But ideally, it feels like we should have 2 different errors at our level if the semantic is important in our business logic. So maybe `DatabaseError` should become an `enum` with one variant being `UniqueViolation`.

Sh099078 replied:
cf https://github.com/OpenRailAssociation/osrd/pull/14198#discussion_r2617394854, done in https://github.com/OpenRailAssociation/osrd/commit/478d629872b8a509a3865b0d5efb82d9fac9480a

leovalais replied:
No need for neither `pub` or parsing once again the violations, it's already done in `Error` in `editoast_models/src/lib.rs`

Sh099078 replied:
Resolving: `add_user_identities` no longer needs to check the nature of the diesel error in its workflow. `DatabaseError` has been reverted to its prior state.

flomonster commented on editoast/editoast_models/src/auth_driver.rs:
```diff
@@ -97,14 +109,31 @@ impl StorageDriver for PgAuthDriver {
+                LEFT JOIN authn_user_identity AS i
+                ON u.id = i.user_id
+                WHERE u.id = $1
+                GROUP BY u.id, u.name;
```
No need to group by name.

```suggestion
                GROUP BY u.id;
```

Khoyo replied:
For those wondering, removing u.name here is a PostgreSQL extension, it's not okay in standard SQL (eg. it would fail with MariaDB). But seeing we have ARRAY_AGG in the same query, we don't really care about standard SQL.

Sh099078 replied:
Fixed in 7b73d8db094dc3b1acebb41edbf3fbd112cc378f

woshilapin commented on editoast/editoast_models/src/auth_driver.rs:
```diff
@@ -242,6 +332,84 @@ impl StorageDriver for PgAuthDriver {
+                                            .execute(&mut conn.write().await)
+                                            .await
+                                            .map_err(|diesel_error| {
+                                                database::DatabaseError::from(diesel_error)
```
It feels weird to wrap and forget about the specificity of the `diesel` error, only for extracting it again 5 lines below because the specificity of the error is finally important to us. See my other comment on the `struct DatabaseError`.

Sh099078 replied:
Agreed, the whole add_user_identities implementation was messy. Refactored it in 478d629872b8a509a3865b0d5efb82d9fac9480a along with the DatabaseError change from struct to enum.

Sh099078 replied:
Thread deprecated by other changes, resolving

woshilapin commented on editoast/editoast_models/src/auth_driver.rs:
```diff
@@ -55,28 +59,36 @@ impl StorageDriver for PgAuthDriver {
+            .inner_join(authn_user::table)
+            .select(authn_user::table::all_columns())
+            .filter(authn_user_identity::identity.eq(user_identity))
+            .first::<(i64, String)>(&mut conn.write().await.deref_mut())
```
Instead of getting `.first` here, why not getting all the lines which contains all the needed identities? It would avoid doing a second request to the database just below that essentially do the same filtering anyway.

Sh099078 replied:
Done in 2fd787247b68f0776444d076ed399644d3e1676d

flomonster reviewed (changes requested):
Something went wrong when testing, I did not investigate much.

When running the migration my single user is duplicated 😬 

**Before:**

 ```
osrd=> select * from authn_user ;
 id | identity_id |     name
----+-------------+--------------
  1 | mock/mocked | Example User
(1 row)
```

**After:**

```
osrd=> select * from authn_user;
 id |     name
----+--------------
  1 | Example User
  2 | mock/mocked
(2 rows)

osrd=> select * from authn_user_identity;
 id | user_id |   identity
----+---------+--------------
  1 |       1 | mock/mocked
  2 |       2 | Example User
```

woshilapin commented on editoast/editoast_models/src/auth_driver.rs:
```diff
@@ -125,36 +154,81 @@ impl StorageDriver for PgAuthDriver {
-                Some(user_info) => {
-                    tracing::debug!(?user_info, "user already exists in db");
-                    Ok(user_info)
+            let mut existing_users = Vec::new();
```
Isn’t it possible to fold into a `HashSet`? It seems that we want to avoid duplicate in this.

Sh099078 replied:
Nice catch. Done in 8dea4ab1b31d4933f5682e453840d22925cc38fc

woshilapin commented on editoast/editoast_models/src/auth_driver.rs:
```diff
@@ -125,36 +154,81 @@ impl StorageDriver for PgAuthDriver {
-
-                None => {
+            }
+            match existing_users.as_slice() {
```
Each branch of the `match` are relatively big, it might help the reading if they were moved into their own function.

Sh099078 replied:
Done in a364cee9890913003d0317a4d15c2641742076b8

woshilapin commented on editoast/editoast_models/src/auth_driver.rs:
```diff
@@ -242,6 +332,84 @@ impl StorageDriver for PgAuthDriver {
+            Some(user_id) => {
+                conn.transaction(async move |conn| {
+                        for new_identity in new_identities {
+                            let values = &vec![(
```
1. Can’t we directly use an array instead of a `Vec` since we know the size at compile-time?

```suggest
                            let values = &[(
```
2. I’m actually surprised that it works, defining a `Vec` then immediately taking the reference. That said, it does look weird and it might be more logical to just have `values` being a `Vec` and use `&value` later 🤷 it’s more of a preference (seems more idiomatic, but maybe only in my head 😅)

Khoyo replied:
For dark magic reasons (aka. *why??*), having an array instead of a slice makes the closure try to borrow `new_identity` as `'static` (the answer probably lies in the trait impl definitions of `Insertable` for `&Vec<T>` and for `&[T]`, but I didn't push that far).

> I’m actually surprised that it works, defining a Vec then immediately taking the reference

That's [temporary lifetime extension](https://doc.rust-lang.org/reference/destructors.html#temporary-lifetime-extension) at work!

Sh099078 replied:
I agree that using an array would be more adequate here but as @Khoyo said, the switching from vector to array is more complex than it should be. I don't think deep diving into how Diesel handles lifetimes internally is worth the effort here: I would rather leave it as is with vectors if you are fine with it @woshilapin

woshilapin commented on editoast/migrations/2025-11-19-092852-0000_modify_authn_user_to_handle_multiple_identities_for_same_user/down.sql:
```diff
@@ -0,0 +1,16 @@
+update authn_user
+set identity_id = identities.identity
+from (
+    select user_id, any_value(identity) as identity
```
I wonder if it’s a good idea to take a non-deterministic function here. Would always taking the first element be a better idea? (I have no idea if Postgres always return the results in the same order for the same query).

Khoyo replied:
> I have no idea if Postgres always return the results in the same order for the same query

Unless there is an `ORDER BY`, you can't rely on any ordering being stable. It might depends on the query plan or table structure changes.

I'd do that if we always want the lowest-id identity

```sql
 select distinct on (user_id) user_id, identity
 from authn_user_identity
 order by user_id, id -- user_id here is required by postgres to do distinct on 
  ```

flomonster replied:
Reverting this migration when users have multiple identities is stupid what so ever. Picking a random identity is fine for me.

Sh099078 replied:
The goal of that update statement is to make it easier for developers to revert the migration to its prior state when they use a dump: the alternatives would have been to completely delete the users or to give them random identities different to the ones they had before the migration. As @flomonster said it doesn't  make any sense on users with multiple identities.

Using `ANY` to select a random identity helps to convey the idea that this is a hack made for development purposes and not a legitimate revert.

I added a comment in the `down.sql` script in 5824dc040f07cee7be91949b74db18fbcbb72f94.

Khoyo replied:
> the alternatives would have been to completely delete the users or to give them random identities different to the ones they had before the migration

No, giving them the lowest id identity should give them the identity they had before the migration, even if you added identities in the meantime. It isn't *that* stupid...

Sh099078 replied:
> > the alternatives would have been to completely delete the users or to give them random identities different to the ones they had before the migration
> 
> No, giving them the lowest id identity should give them the identity they had before the migration, even if you added identities in the meantime. It isn't _that_ stupid...

On second thought I agree with you, we might aswell keep the same identity they had before the migration.

Update statement changed in c8a5bba6903ac3f8a59195f3a24ec3f0cafb0f62 to keep the oldest identity of each user instead of picking it at random.

woshilapin commented on editoast/src/views/mod.rs:
```diff
@@ -618,26 +618,31 @@ async fn authenticate(
-            Err(err) => return Err(err.into()),
-        };
+    let authorizer = match Authorizer::try_initialize(
+        identity.clone().expect("Already checked above"),
```
Perhaps the code above should return a tuple of `(user, identity)` instead of just the `user` then? This way, I believe the `expect()` won’t be needed.

Sh099078 replied:
Indeed, that's better. Done in bb85447d7651b1f8e2de197d4c8dca31f2c1cabc

woshilapin reviewed (commented):
It would have been nice to separate the database model update from the add of the CLI `add-identities` as both add significant number of lines of code and I believe, they could have been review independently (maybe even in a different PR).

Sh099078 commented:
> Something went wrong when testing, I did not investigate much.
> 
> When running the migration my single user is duplicated 😬
> 
> **Before:**
> 
> ```
> osrd=> select * from authn_user ;
> id | identity_id |     name
> ----+-------------+--------------
>  1 | mock/mocked | Example User
> (1 row)
> ```
> 
> **After:**
> 
> ```
> osrd=> select * from authn_user;
>  id |     name
> ----+--------------
>   1 | Example User
>   2 | mock/mocked
> (2 rows)
> 
> osrd=> select * from authn_user_identity;
>  id | user_id |   identity
> ----+---------+--------------
>   1 |       1 | mock/mocked
>   2 |       2 | Example User
> ```

Was that on the latest commit ? The CLI parameters order got reversed in the CLI subcommand `user add` and went from `user add [identity] [name]` to `user add [name] [identities]`. I fixed the docker compose quite recently: if you ran it on an older version they might still appear in the wrong order. If so, starting the pods will first run the diesel migration and correctly update the existing "Example User" but then the call to `add user` will create another example user with its name and identity reversed.

Testing on my side, I ran the migration twice: the first time directly with the Diesel CLI and the second time indirectly by starting the containers. I got the expected result in both cases:

```
osrd=# select * from authn_user;
 id |     name
----+--------------
  1 | Example User
(1 row)

osrd=# select * from authn_user_identity;
 id | user_id |  identity
----+---------+-------------
  1 |       1 | mock/mocked
(1 row)
```

Sh099078 committed:
editoast, front: remove user identity from search request and response

On the front-end side, use the endpoint `authz/user/info` to retrieve
the user identity instead of getting it from a search query.

Signed-off-by: Loup Federico <16464925+Sh099078@users.noreply.github.com>

Sh099078 committed:
editoast: support multiple identities per user

- Update the database model to be able to associate several identities
  per user.
- Add a CLI subcommand `user add-identities` to add additional identities
  to an existing user.

Signed-off-by: Loup Federico <16464925+Sh099078@users.noreply.github.com>

gaetan-osrd commented on docker-compose.yml:
```diff
@@ -147,7 +147,7 @@ services:
         fga_migrations apply &&
         diesel migration run --locked-schema &&
-        editoast user add 'mock/mocked' 'Example User' &&
+        editoast user add 'Example User' 'mock/mocked' &&
```
Should be warned a s a breaking change in the release for external users who might avec some automation. An affected user is demo.osrd.fr for example

Sh099078 replied:
I pinged you in order to see what should where we should put those warnings. I don't think it requires any change in the PR itself so I am resolving that thread: feel free to re-open it I was mistaken in doing so.

aschantraine commented on editoast/src/client/user.rs:
```diff
@@ -25,6 +26,22 @@ pub enum UserCommand {
     Info(InfoArgs),
     /// Delete a user
     Delete(DeleteArgs),
+    /// Add an identity to a new or an already existing user
```
Suggestion : ` /// Add identities to a new or an already existing user` as multiple identities can be added at once.

leovalais commented on editoast/editoast_models/src/auth_driver.rs:
```diff
@@ -55,28 +58,36 @@ impl StorageDriver for PgAuthDriver {
-            .select(authn_user::all_columns)
-            .filter(authn_user::identity_id.eq(&user_identity))
-            .first::<(i64, String, String)>(conn.write().await.deref_mut())
+        let related_user = authn_user_identity::table
```
transaction

Sh099078 replied:
Deprecated by https://github.com/OpenRailAssociation/osrd/commit/2fd787247b68f0776444d076ed399644d3e1676d: there is only one postgres query now (https://github.com/OpenRailAssociation/osrd/pull/14198#discussion_r2613612685).

Resolving that thread

leovalais commented on editoast/editoast_models/src/auth_driver.rs:
```diff
@@ -125,36 +153,81 @@ impl StorageDriver for PgAuthDriver {
     async fn ensure_user(&self, user: &UserInfo) -> Result<User, Self::Error> {
```
- Take only a name and a single identity => no edge cases
- Nothing to change in the middleware
- In the CLI: ensure with the first identity and append the rest in a second step

Sh099078 replied:
Done in 55db9c3f3947d4a0354363db732bb1c2b0cc20af

leovalais commented on editoast/editoast_models/src/auth_driver.rs:
```diff
@@ -205,11 +278,27 @@ impl StorageDriver for PgAuthDriver {
+                    diesel::sql_types::Array<
+                        diesel::sql_types::Nullable<diesel::sql_types::Varchar>,
+                    >,
+                >("array_agg(identity)"),
```
`sql::<...>("array_agg(").bind(authn_user_identity::identity).sql(")")`

So as to not hardcode the name of the column (may or may not work...)

Sh099078 replied:
It works. Done in dc786316b85c209d6f1b2049066cd0950da9fc56

leovalais commented on editoast/editoast_models/src/auth_driver.rs:
```diff
@@ -205,11 +278,27 @@ impl StorageDriver for PgAuthDriver {
+            ))
+            .load_stream::<(i64, String, Vec<Option<String>>)>(&mut conn.write().await)
             .await?
             .map(|res| match res {
```
try `map_ok` from `TryStreamExt`

leovalais commented on editoast/migrations/2025-11-19-092852-0000_modify_authn_user_to_handle_multiple_identities_for_same_user/up.sql:
```diff
@@ -0,0 +1,11 @@
+create table authn_user_identity (
```
Maybe create a few triggers to ensure the consistency of the model in the DB itself to avoid relying on prior checks that may break.

- on insert on authz_user: DEFERRED, make sure an identity is set before the end of the transaction
- on update: check consistency at the end as well
- on delete on authn_user_identity: remove the user if the last identity is removed

Sh099078 replied:
Done in 3653cab85f8870d1c67d0cd07deacab14c0348c1 along with the addition of some indexes in `authn_user` and fixes in the `down.sql script` drop statements.

Note on the drop statements: it seems our migrations make an extensive use of `IF EXISTS` clauses when creating or deleting database objects. Open to discussion: I think that this is a bad habit that potentially hides mistakes in the migrations and doesn't give anything valuable in return and that we should stop doing it in future migrations.
- On objects deletion, it hides typos: `DROP INDEX IF EXISTS index_name_with_typo` will not crash when running a migration script even if the given index name is incorrect. We might end up with incorrect reverts scripts that miss the deletion of some indexes, functions, etc by generalizing the use of `IF EXISTS`.
- On objects creation, it hides invalid database states: when creating a function or an index for the first time, we want the migration to fail if a database object already exists if it should not have been created yet, as it probably means that the migrations revert script does not properly reset the database to its prior state (drop statement missing ?).

leovalais replied:
These `IF [NOT ]EXISTS` are useful when applying/redoing/fixing failed migrations while developing. But they shouldn't theoretically remain in the final migration file, agreed.

leovalais commented on editoast/src/client/user.rs:
```diff
@@ -36,10 +36,10 @@ pub struct ListArgs {
-    identity: String,
     /// Name of the user
     name: Option<String>,
+    /// Identity of the user
```
Plural

leovalais commented on editoast/src/views/authz.rs:
```diff
@@ -17,6 +18,11 @@ use axum::http::StatusCode;
+use diesel::ExpressionMethods;
+use diesel::QueryDsl;
+use diesel::prelude::*;
+use diesel_async::RunQueryDsl;
```
Make functions in editoast_models

Sh099078 replied:
Done in e2d7615bcfa2c7b4ad02ea8699b0afecf81baff8

leovalais commented on editoast/src/views/test_app.rs:
```diff
@@ -345,11 +345,12 @@ impl TestApp {
         &self.app_state.config
     }
 
+    // TODO PR allow the creation of users with multiple identities for the tests ?
```
Maybe remove the comment, if someone needs such function they'll write it

leovalais commented on editoast/src/client/user.rs:
```diff
@@ -25,6 +26,22 @@ pub enum UserCommand {
+    pub user_id: Option<i64>,
+    #[arg(short, long)]
+    pub identity: Option<String>,
+    pub identities: Vec<String>,
```
pub?

Sh099078 replied:
oops. Fixed in a3081c4b87f3f7d8c263d67e2be5fa25e476480b

leovalais commented on editoast/src/client/user.rs:
```diff
@@ -25,6 +26,22 @@ pub enum UserCommand {
+    pub user_id: Option<i64>,
+    #[arg(short, long)]
+    pub identity: Option<String>,
+    pub identities: Vec<String>,
```
rename: user_identity and new_identities

leovalais commented on editoast/src/client/user.rs:
```diff
@@ -170,3 +187,24 @@ pub async fn delete_user(
+    let user_identity = match (user_id, identity) {
+        (Some(user_id), _) => Either::Left(user_id),
+        (_, Some(identity)) => Either::Right(identity.clone()),
+        _ => panic!("No user_id or identity provided"),
```
unreachable!("ensured by clap")

leovalais commented on editoast/src/client/user.rs:
```diff
@@ -170,3 +187,24 @@ pub async fn delete_user(
+    let driver = PgAuthDriver::new(pool);
+    let user_identity = match (user_id, identity) {
+        (Some(user_id), _) => Either::Left(user_id),
+        (_, Some(identity)) => Either::Right(identity.clone()),
```
get the ID here from the identity and just pass it to add_user_identities instead of an Either

Sh099078 replied:
Summary of a discussion with @leovalais:

tl;dr: that change is more annoying than expected, lets not do this.

Short explanation:
- If we get the id from the identity (or the opposite) in `client::user::add_identities`, we need to either do it in a postgres query outside of the transaction started in `PgAuthDriver::add_user_identities`, or somehow start the transaction in `client::user::add_identities` and make it available to `PgAuthDriver::add_user_identities` (which might require some bigger refactoring out of scope for this PR).
- Removing either the identity or the identifier from the CLI parameters is not an option because both are likely to be used by ops.

=> Passing an Either<i64, String> from the client to `PgAuthDriver::add_user_identities` remains the simplest way to keep everything in the same transaction. We are keeping it this way for the time being.

Resolving that thread

leovalais commented on editoast/authz/src/lib.rs:
```diff
@@ -342,6 +343,28 @@ mod mock_driver {
+            match user_id {
+                Some(user_id) => new_identities.iter().for_each(|new_identity| {
+                    map.insert(new_identity.to_string(), user_id);
+                }),
```
nit: you can probably do:

```rust
map.extend(std::iter::repeat(user_id).zip(new_identities.iter().map(ToString::to_string)))
```

leovalais commented on editoast/editoast_models/src/auth_driver.rs:
```diff
@@ -331,6 +332,84 @@ impl StorageDriver for PgAuthDriver {
     }
 
+    #[tracing::instrument(skip_all, fields(identities, user = %user_identity), ret(level = Level::DEBUG), err)]
+    async fn add_user_identities(
```
We're only using it in the CLI, no need for batching, we can simply do:

- take one identity
- open transacton
- first query to check that the identity doesn't exist
- insert the identity for the user

Should simplify the workflow, remove edge cases and make error manager much simpler.

leovalais reviewed (commented):
Peer reviewed with @Sh099078

aschantraine commented on editoast/editoast_models/src/auth_driver.rs:
```diff
@@ -262,6 +378,84 @@ impl StorageDriver for PgAuthDriver {
+        Ok(User {
+            id: existing_user.id,
+            info: UserInfo {
+                name: user.name.to_owned(),
```
The name is not updated if it doesn't match the existing one so we should take the name from `existing_user` here.
Use case :
`cargo run user add "u_1" "i_1"` => "User added with id: u_1 (i_1) [201]"
`cargo run user add "u_2" "i_1"` => "User added with id: u_2 (i_1) [201]" (should be u_1 since the name stays u_1 (current behavior))

We could also choose to raise an error in this case but the current behavior is as described above.

Sh099078 replied:
Nice catch! Refactoring `ensure_user` following https://github.com/OpenRailAssociation/osrd/pull/14198#discussion_r2619761191 fixed this issue. The current behaviour is to fail in such case. Resolving the tread

aschantraine reviewed (commented):
Tested and seems to work well (minor details)

Synar commented on front/src/common/api/baseGeneratedApis.ts:
```diff
@@ -28,7 +28,7 @@ export const baseEditoastApi = createApi({
 
       if (impersonatedUser) {
-        headers.set('x-impersonate', impersonatedUser.identity_id);
+        headers.set('x-impersonate', impersonatedUser.identities[0]);
```
Dumb question, but does the model guarantees that a user has at least one identity? I assume yes, but just checking.

Sh099078 replied:
It is part of the todo-list: some comments in @leovalais review are about enforcing it through database triggers. At the moment there is no way to create a user with no identity or to delete user identities so they *should* always have at least one but only enforcing it through the code isn't great indeed.

Khoyo replied:
In any case, trying to impersonate such a user is impossible for the front.

It might be a good idea to be a little defensive and reject to impersonate such a user in impersonate ?

Sh099078 replied:
The database model has been updated to enforce that users have at least one identity through triggers so it should never happen (unless there is a bug in the way Editoast retrieves user identities).

6ea40ba17d05e41a0a9b82052a3ff5afa193797b adds a front-end check that the identities list returned by Editoast is not empty added.

Synar commented on front/src/utils/types.ts:
```diff
@@ -20,6 +22,17 @@ export const mapBy = <K extends keyof T, T>(items: T[] | undefined, key: K): Map
 export const concatMap = <K, V>(map1: Map<K, V>, map2: Map<K, V>) =>
   new Map([...map1.entries(), ...map2.entries()]);
 
+export type UserInfo = {
```
Not completely sure it's the best place for the type, but it makes some sense. @ maintainer-that-will-review-this what do you think?

clarani replied:
Yes 100% agree, this file contains utils functions about types.

Since it is only used in `reducer/user/index.ts`, it would be better to move it there directly

Sh099078 replied:
Done in e185b8386fc3a5331967520d88ca4da8f5833f0d

Synar reviewed (approved):
LGTM for the front part ^^

flomonster commented:
> Was that on the latest commit ? The CLI parameters order got reversed in the CLI subcommand `user add` and went from `user add [identity] [name]` to `user add [name] [identities]`. I fixed the docker compose quite recently: if you ran it on an older version they might still appear in the wrong order. If so, starting the pods will first run the diesel migration and correctly update the existing "Example User" but then the call to `add user` will create another example user with its name and identity reversed.
> 
> Testing on my side, I ran the migration twice: the first time directly with the Diesel CLI and the second time indirectly by starting the containers. I got the expected result in both cases:
> 
> ```
> osrd=# select * from authn_user;
>  id |     name
> ----+--------------
>   1 | Example User
> (1 row)
> 
> osrd=# select * from authn_user_identity;
>  id | user_id |  identity
> ----+---------+-------------
>   1 |       1 | mock/mocked
> (1 row)
> ```

Weird, I retested it on the latest commit of this PR using `./scripts/pr-tests-compose.sh` and the latest dump. Maybe it's another issue not linked to this PR.

If I'm the only one with this issue, let's not block the merge of this PR.

Khoyo commented on editoast/migrations/2025-11-19-092852-0000_modify_authn_user_to_handle_multiple_identities_for_same_user/up.sql:
```diff
@@ -0,0 +1,57 @@
+create table authn_user_identity (
+    id bigint primary key generated by default as identity,
+    user_id bigint references authn_user on delete cascade not null,
```
There needs to be an index on this column (to avoid a full table scan when deleting a user or listing identities for a user)

leovalais replied:
I keep forgetting about this, so annoying it's not a default

Sh099078 replied:
Index added in 01d3d0c43322931b0a8079ea29f1867f86b4f7f1, resolving thread

flomonster reviewed (approved):
LGTM, just need to fix remaining comments.

leovalais commented on editoast/editoast_models/src/auth_driver.rs:
```diff
@@ -97,14 +123,31 @@ impl StorageDriver for PgAuthDriver {
+                ON u.id = i.user_id
+                WHERE u.id = $1
+                GROUP BY u.id;
+            ";
```
can't we deduplicate this with `get_batch_user_identities`?

Sh099078 replied:
We don't need exactly the information it returns. `get_batch_user_identities` returns user identifiers and their identities list, here we need the user names and their identities. I can update `get_batch_user_identities` to also return usernames ?

leovalais replied:
> I can update get_batch_user_identities to also return usernames

Yes, we can waste a little memory to deduplicate a raw sql request

Sh099078 replied:
Done in f86bd1659974a7f575574865b850ba14977953e7

leovalais commented on editoast/src/client/user.rs:
```diff
@@ -170,3 +190,24 @@ pub async fn delete_user(
+    let user_identity = match (user_id, user_identity) {
+        (Some(user_id), _) => Either::Left(user_id),
+        (_, Some(identity)) => Either::Right(identity.clone()),
+        _ => unreachable!("No user_id or identity provided"),
```
the string is supposed to be the reason **why** this is unreachable (just like expect): this looks like a user error

Sh099078 replied:
My bad, I forgot to update the description when changing for `unreachable!`. Fixed in f20ed81097777e165b69f72d518acf772ebb36fb

leovalais commented on editoast/editoast_models/src/auth_driver.rs:
```diff
@@ -262,6 +388,46 @@ impl StorageDriver for PgAuthDriver {
 }
 
+impl PgAuthDriver {
+    async fn save_new_user(
```
why not use `ensure_user`?

Sh099078 replied:
`save_new_user` is called in `ensure_user`, it is the function that does the database persistence

leovalais replied:
oh my bad

leovalais commented on editoast/editoast_models/src/authn/user.rs:
```diff
@@ -7,6 +19,68 @@ use crate as editoast_models; // HACK: remove after all models are in this crate
+    pub async fn get_batch_user_identities(
+        user_ids: &[i64],
+        conn: &mut DbConnection,
+    ) -> Result<HashMap<i64, (UserName, Vec<UserIdentity>)>, diesel::result::Error> {
```
```suggestion
    ) -> Result<HashMap<i64, UserInfo>, diesel::result::Error> {
```
